Add WKTouchEventsGestureRecognizer and use it to replace UIWebTouchEventsGestureRecognizer

whsieh opened the pull request:
#### 153d88fa3ea612b1b3db28834cc3232a8b5c812e
<pre>
Add WKTouchEventsGestureRecognizer and use it to replace UIWebTouchEventsGestureRecognizer
<a href="https://bugs.webkit.org/show_bug.cgi?id=258748">https://bugs.webkit.org/show_bug.cgi?id=258748</a>
rdar://111922409

Reviewed by Megan Gardner.

Introduce `WKTouchEventsGestureRecognizer`, and use it to replace `UIWebTouchEventsGestureRecognizer`
from UIKit. This allows us to observe and dispatch touch events, using only UIKit API.

No change in behavior.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/Shared/NativeWebTouchEvent.h:
* Source/WebKit/Shared/ios/NativeWebTouchEventIOS.mm:
(WebKit::webEventTypeForWKTouchEventType):
(WebKit::convertTouchType):
(WebKit::radiusForTouchPoint):
(WebKit::NativeWebTouchEvent::extractWebTouchPoint):
(WebKit::NativeWebTouchEvent::NativeWebTouchEvent):
(WebKit::webEventTypeForUIWebTouchEventType): Deleted.
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::doneWithTouchEvent):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setUpInteraction]):
(-[WKContentView cleanUpInteraction]):
(-[WKContentView _touchEventsRecognized:]):
(-[WKContentView _touchEvent:preventsNativeGestures:]):
(-[WKContentView touchEventGestureRecognizer]):
(-[WKContentView gestureRecognizer:shouldIgnoreTouchEvent:]):
(-[WKContentView _webTouchEventsRecognized:]): Deleted.
(-[WKContentView _webTouchEvent:preventsNativeGestures:]): Deleted.

Also take the opportunity to simplify some of the method names here by removing the &quot;web&quot; prefix in
cases where it was only necessary to differentiate WebKit-specific code from other UIKit classes.

(-[WKContentView gestureRecognizer:shouldIgnoreWebTouchWithEvent:]): Deleted.
* Source/WebKit/UIProcess/ios/WKSyntheticTapGestureRecognizer.h:
* Source/WebKit/UIProcess/ios/WKSyntheticTapGestureRecognizer.mm:
(-[WKSyntheticTapGestureRecognizer touchesEnded:withEvent:]):
* Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizer.h: Added.
* Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizer.mm: Added.
(approximateWallTime):
(+[WKTouchEventsGestureRecognizer initialize]):
(-[WKTouchEventsGestureRecognizer initWithTarget:action:touchDelegate:]):
(-[WKTouchEventsGestureRecognizer activeTouchesByIdentifier]):
(-[WKTouchEventsGestureRecognizer reset]):
(-[WKTouchEventsGestureRecognizer cancel]):
(nextTouchIdentifier):
(-[WKTouchEventsGestureRecognizer _updateTapStateWithTouches:]):
(-[WKTouchEventsGestureRecognizer _updateTapStateWithTouches:type:]):
(-[WKTouchEventsGestureRecognizer _recordTouches:type:]):
(eventTypeForTouchPhase):
(lastExpectedWKEventTypeForTouches):
(-[WKTouchEventsGestureRecognizer performAction]):
(-[WKTouchEventsGestureRecognizer _hasActiveTouchesForEvent:]):
(-[WKTouchEventsGestureRecognizer _processTouches:withEvent:type:]):
(-[WKTouchEventsGestureRecognizer canBePreventedByGestureRecognizer:]):
(-[WKTouchEventsGestureRecognizer touchesBegan:withEvent:]):
(-[WKTouchEventsGestureRecognizer touchesMoved:withEvent:]):
(-[WKTouchEventsGestureRecognizer touchesEnded:withEvent:]):
(-[WKTouchEventsGestureRecognizer touchesCancelled:withEvent:]):
(-[WKTouchEventsGestureRecognizer lastTouchEvent]):
* Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizerTypes.h: Copied from Source/WebKit/UIProcess/ios/WKSyntheticTapGestureRecognizer.h.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Configurations/Base.xcconfig:
* Tools/TestWebKitAPI/Tests/ios/TouchEventTests.mm:
(touchEventsGestureRecognizerClass):
(-[WKWebView touchEventGestureRecognizer]):
(TestWebKitAPI::updateSimulatedTouchEvent):
(TestWebKitAPI::simulatedTouchEvent):
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/ios/UIKitSPI.h:

Canonical link: <a href="https://commits.webkit.org/265916@main">https://commits.webkit.org/265916@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c42f52c20ed5eb86492912f2bce2c0c69b035f7d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12251 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12596 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12899 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13996 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11819 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12281 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15014 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12612 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14494 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12415 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13205 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10375 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14416 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10490 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11116 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11571 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11280 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11775 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/9714 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10991 "Built successfully") | | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3013 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15322 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11630 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizer.h:
```diff
@@ -0,0 +1,84 @@
+- (BOOL)isAnyTouchOverActiveArea:(NSSet *)touches;
+@optional
+- (BOOL)shouldIgnoreWebTouch;
+- (BOOL)gestureRecognizer:(WKTouchEventsGestureRecognizer *)gestureRecognizer shouldIgnoreWebTouchWithEvent:(UIEvent *)event;
```
What's the difference between a "web touch" and a normal touch?

whsieh replied:
~What's a "normal touch"?~

(Edit: I see, is your suggestion to rename some uses of `WebTouch` to just `Touch`? IIUC, that seems reasonable to me).

aprotyas replied:
Yeah I think `WebTouch` is just `Touch` but in WebKit. Maybe it helped differentiate from `WebCore::Touch` at a time

megangardner commented on Source/WebKit/Shared/ios/NativeWebTouchEventIOS.mm:
```diff
@@ -113,26 +107,25 @@ static CGFloat radiusForTouchPoint(const _UIWebTouchPoint& touchPoint)
         platformTouchPoint.setRadiusX(radius);
         platformTouchPoint.setRadiusY(radius);
-        platformTouchPoint.setRotationAngle(0); // Not available in _UIWebTouchEvent yet.
+        platformTouchPoint.setRotationAngle(0); // Not available in WKTouchPoint yet.
```
Can we add this? If so, can we make a radar and a FIXME: instead?

whsieh replied:
I filed [a new bug](https://webkit.org/b/259068) to implement this, which I'll reference in this comment with a FIXME (I avoided adding it here, since this patch should just preserve existing behavior).

megangardner commented on Source/WebKit/Shared/ios/NativeWebTouchEventIOS.mm:
```diff
@@ -113,26 +107,25 @@ static CGFloat radiusForTouchPoint(const _UIWebTouchPoint& touchPoint)
         auto radius = radiusForTouchPoint(touchPoint);
         platformTouchPoint.setRadiusX(radius);
```
This feels like really dumb questions, but where is platformTouchPoint defined?? I don't see it.

whsieh replied:
It's defined a few lines above ([NativeWebTouchEventIOS.mm:105](https://github.com/WebKit/WebKit/pull/15548/files#diff-2fcda1cf3dd4160b0b33c0a6213b5337d3bb4168587ead6c18486e4e75e07751R105))

whsieh committed:
Add WKTouchEventsGestureRecognizer and use it to replace UIWebTouchEventsGestureRecognizer
https://bugs.webkit.org/show_bug.cgi?id=258748
rdar://111922409

Reviewed by Megan Gardner.

Introduce `WKTouchEventsGestureRecognizer`, and use it to replace `UIWebTouchEventsGestureRecognizer`
from UIKit. This allows us to observe and dispatch touch events, using only UIKit API.

No change in behavior.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/Shared/NativeWebTouchEvent.h:
* Source/WebKit/Shared/ios/NativeWebTouchEventIOS.mm:
(WebKit::webEventTypeForWKTouchEventType):
(WebKit::convertTouchType):
(WebKit::radiusForTouchPoint):
(WebKit::NativeWebTouchEvent::extractWebTouchPoint):
(WebKit::NativeWebTouchEvent::NativeWebTouchEvent):
(WebKit::webEventTypeForUIWebTouchEventType): Deleted.
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::doneWithTouchEvent):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setUpInteraction]):
(-[WKContentView cleanUpInteraction]):
(-[WKContentView _touchEventsRecognized:]):
(-[WKContentView _touchEvent:preventsNativeGestures:]):
(-[WKContentView touchEventGestureRecognizer]):
(-[WKContentView gestureRecognizer:shouldIgnoreTouchEvent:]):
(-[WKContentView _webTouchEventsRecognized:]): Deleted.
(-[WKContentView _webTouchEvent:preventsNativeGestures:]): Deleted.

Also take the opportunity to simplify some of the method names here by removing the "web" prefix in
cases where it was only necessary to differentiate WebKit-specific code from other UIKit classes.

(-[WKContentView gestureRecognizer:shouldIgnoreWebTouchWithEvent:]): Deleted.
* Source/WebKit/UIProcess/ios/WKSyntheticTapGestureRecognizer.h:
* Source/WebKit/UIProcess/ios/WKSyntheticTapGestureRecognizer.mm:
(-[WKSyntheticTapGestureRecognizer touchesEnded:withEvent:]):
* Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizer.h: Added.
* Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizer.mm: Added.
(approximateWallTime):
(+[WKTouchEventsGestureRecognizer initialize]):
(-[WKTouchEventsGestureRecognizer initWithTarget:action:touchDelegate:]):
(-[WKTouchEventsGestureRecognizer activeTouchesByIdentifier]):
(-[WKTouchEventsGestureRecognizer reset]):
(-[WKTouchEventsGestureRecognizer cancel]):
(nextTouchIdentifier):
(-[WKTouchEventsGestureRecognizer _updateTapStateWithTouches:]):
(-[WKTouchEventsGestureRecognizer _updateTapStateWithTouches:type:]):
(-[WKTouchEventsGestureRecognizer _recordTouches:type:]):
(eventTypeForTouchPhase):
(lastExpectedWKEventTypeForTouches):
(-[WKTouchEventsGestureRecognizer performAction]):
(-[WKTouchEventsGestureRecognizer _hasActiveTouchesForEvent:]):
(-[WKTouchEventsGestureRecognizer _processTouches:withEvent:type:]):
(-[WKTouchEventsGestureRecognizer canBePreventedByGestureRecognizer:]):
(-[WKTouchEventsGestureRecognizer touchesBegan:withEvent:]):
(-[WKTouchEventsGestureRecognizer touchesMoved:withEvent:]):
(-[WKTouchEventsGestureRecognizer touchesEnded:withEvent:]):
(-[WKTouchEventsGestureRecognizer touchesCancelled:withEvent:]):
(-[WKTouchEventsGestureRecognizer lastTouchEvent]):
* Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizerTypes.h: Copied from Source/WebKit/UIProcess/ios/WKSyntheticTapGestureRecognizer.h.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Configurations/Base.xcconfig:
* Tools/TestWebKitAPI/Tests/ios/TouchEventTests.mm:
(touchEventsGestureRecognizerClass):
(-[WKWebView touchEventGestureRecognizer]):
(TestWebKitAPI::updateSimulatedTouchEvent):
(TestWebKitAPI::simulatedTouchEvent):
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/ios/UIKitSPI.h:

Canonical link: https://commits.webkit.org/265916@main
